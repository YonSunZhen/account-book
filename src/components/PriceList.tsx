import React from 'react';
import Ionicon from 'react-ionicons';
import ProTypes from 'prop-types';

interface PriceListProp {
  items?: any,
  onModifyItem?: Function;
  onDeleteItem?: Function;
}

const PriceList = ({ items = [], onModifyItem = () => {}, onDeleteItem = () => {} }: PriceListProp) => {
  return (
    <ul className='list-group list-grop-flush'>
      {
        items.map((item) => (
          <li className='list-group-item d-flex justify-content-between align-items-center' key={item.id}>
            <span className='col-1 badge badge-primary'>
              <Ionicon
                className='rounded-circle'
                fontSize='30px'
                style={{backgroundColor: '#007bff', padding: '5px'}}
                color={'#fff'}
                icon={item.category.iconName}
              />
            </span>
            <span className='col-5'>{item.title}</span>
            <span className='col-2 font-weight-bold'>
              {(item.category.type === 'income') ? '+' : '-'}{item.price}å…ƒ
            </span>
            <span className='col-2'>{item.date}</span>
            <a className='col-1' href="" onClick={() => {onModifyItem(item);}}>
              <Ionicon
                className='rounded-circle'
                fontSize='30px'
                style={{backgroundColor: '#28a745', padding: '5px'}}
                color={'#fff'}
                icon='ios-create-outline'
              />
            </a>
            <a className='col-1' onClick={() => {onDeleteItem(item);}}>
              <Ionicon
                className='rounded-circle'
                fontSize='30px'
                style={{backgroundColor: '#dc3545', padding: '5px'}}
                color={'#fff'}
                icon='ios-close'
              />
            </a>
          </li>
        ))
      }
    </ul>
  );
};

// PriceList.propTypes = {
//   items: ProTypes.array.isRequired,
//   onModifyItem: ProTypes.func.isRequired,
//   onDeleteItem: ProTypes.func.isRequired
// }

export default PriceList;