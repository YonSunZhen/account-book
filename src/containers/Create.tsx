import React from 'react';
import CategorySelect from '../components/CategorySelect';
import PriceForm from '../components/PriceForm';
import Tabs, { Tab } from '../components/Tabs';
import { testCategories } from '../testData';
import { TYPE_INCOME, TYPE_OUTCOME } from '../utility';

interface Prop {

}

interface State {

}
class Create extends React.Component<Prop, State> {

  onSelectCategoryChange = (category) => {  
  }

  render() {
    const filterCategories = testCategories.filter(category => category.type === TYPE_OUTCOME);
    return (
      <div className=''>
        <Tabs activeIndex={0} onTabChange={() => {}}>
          <Tab>支出</Tab>
          <Tab>收入</Tab>
        </Tabs>
        <CategorySelect categories={filterCategories} onSelectCategory={this.onSelectCategoryChange}></CategorySelect>
        <PriceForm></PriceForm>
      </div>
    );
  }
}

export default Create;