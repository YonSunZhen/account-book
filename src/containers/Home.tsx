import React, { Component } from 'react';
import MonthPicker from '../components/MonthPicker';
import {LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, padLeft} from '../utility';
import PriceList from '../components/PriceList';
import Tabs, { Tab } from '../components/Tabs';
import CreateBtn from '../components/CreateBtn';
import TotalPrice from '../components/TotalPrice';
import Day from 'dayjs';
import Ionicon from 'react-ionicons';
import withContext from '../WithContext';
import { RouterProps } from 'react-router-dom';

interface Prop extends RouterProps {
  data?: any;
}

interface State {
  items?: Item[];
  selectedYear?: string;
  selectedMonth?: string;
  tabView?: string;
}

interface Item {
  id?: number;
  title?: string;
  price?: number;
  date?: string;
  cid: number;
  categroy?: any;
}

export const categories = {
  '1': {
    name: '旅行',
    type: 'outcome',
    iconName: 'ios-plane'
  },
  '2': {
    name: '理财收入',
    type: 'income',
    iconName: 'ios-plane'
  }
};

export const items: Item[] = [
  {
    id: 1,
    title: '去云南旅游',
    price: 200,
    date: '2018-08-10',
    cid: 1
  },
  {
    id: 1,
    title: '去云南旅游',
    price: 200,
    date: '2018-09-10',
    cid: 1
  },
  {
    id: 1,
    title: '天天基金',
    price: 200,
    date: '2018-09-10',
    cid: 2
  }
];

export const newItem = {
  id: 1,
  title: '天天基金',
  price: 200,
  date: '2018-09-10',
  cid: 2
};

const tabViewList = [LIST_VIEW, CHART_VIEW];

class Home extends Component<Prop, State> {

  constructor(props: Prop) {
    super(props);
    this.state = {
      items,
      selectedYear: String(Day().year(2018).format('YYYY')),
      selectedMonth: String(Day().month(8).format('M')),
      tabView: LIST_VIEW
    };
  }

  // 必须使用箭头函数
  onChangeDate = (year, month) => {
    this.setState({
      selectedYear: year,
      selectedMonth: month
    });
  }

  onChangeView = (index) => {
    this.setState({
      tabView: tabViewList[index]
    });
  }

  onCreateItem = () => {
    this.props.history.push('/create');
  }

  onModifyItem = (modifiedItem) => {
    this.props.history.push(`/edit/${modifiedItem.id}`);
  }

  onDelItem = (deletedItem) => {
    const filterdItems = this.state.items?.filter(item => item.id !== deletedItem);
    this.setState({
      items: filterdItems
    });
  }

  render() {
    const { data } = this.props;
    const { items = [], selectedYear, selectedMonth, tabView } = this.state;
    const itemsWithCategroy = items.map(item => {
      item.categroy = categories[item.cid];
      return item;
    }).filter(item => item.date?.includes(`${selectedYear}-${padLeft(selectedMonth)}`));
    let totalIncome = 0; 
    let totalOutcome = 0;
    itemsWithCategroy.forEach(item => {
      if (item.categroy.type === TYPE_OUTCOME) {
        totalOutcome += item.price || 0;
      } else {
        totalIncome += item.price || 0;
      }
    });
    return (
      <React.Fragment>
        <div className='App-header'>
          <div className='row'>
            <div className='col'>
              <MonthPicker
                year={this.state.selectedYear}
                month={this.state.selectedMonth}
                onChange={this.onChangeDate}
              />
            </div>
            <div className='col'>
              <TotalPrice
                income={totalIncome}
                outcome={totalOutcome}
              />
            </div>
          </div>
        </div>
        <div className='content-area py3 px-3'>
          <Tabs activeIndex={0} onTabChange={this.onChangeView}>
            <Tab>
              <Ionicon
                className='rounded-circle mr-z'
                fontSize='25px'
                color={'#007bff'}
                icon='ios-paper'
              />
              列表模式
            </Tab>
            <Tab>
              <Ionicon
                className='rounded-circle mr-z'
                fontSize='25px'
                color={'#007bff'}
                icon='ios-pie'
              />
              图表模式
            </Tab>
          </Tabs>
          <CreateBtn onClick={this.onCreateItem}></CreateBtn>
          { tabView === LIST_VIEW &&
            <PriceList 
              items={itemsWithCategroy} 
              onDeleteItem={this.onDelItem}
              onModifyItem={this.onModifyItem}
            />
          }
          { tabView === CHART_VIEW &&
            <h1 className='chart-title'>图表区域</h1>
          }
        </div>
      </React.Fragment>
    );
  }
}

export default withContext(Home);