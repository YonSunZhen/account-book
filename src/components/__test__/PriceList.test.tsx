import { shallow } from 'enzyme';
import PriceList from '../PriceList';
import Ionicon from 'react-ionicons';
import { items, categories } from '../../containers/Home';

const itemsWithCategroy = items.map(item => {
  item.categroy = categories[item.cid];
  return item;
});

const props = {
  items: itemsWithCategroy,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn()
};
let wrapper;
describe('test PriceList component', () => {
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props}></PriceList>);
  });

  it('should render the component to match snapshots', () => {
    expect(wrapper).toMatchSnapshot(); // FIXME: 不生效
  });

  it('should render correct price items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategroy.length);
  });

  it('should render correct icon and price for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon);
    expect(iconList.length).toEqual(3);
    // expect(iconList.first().props().icon).toEqual(itemsWithCategroy[0].categroy.iconName);
  });

  it('should trigger the correct function callbacks', () => {
    const firstItem = wrapper.find('.list-group-item').first();
    firstItem.find('a').first().simulate('click'); // 模拟点击事件
    expect(props.onModifyItem).toHaveBeenCalledWith(itemsWithCategroy[0]); // 触发事件响应
    firstItem.find('a').last().simulate('click');
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategroy[0]);
  });
});