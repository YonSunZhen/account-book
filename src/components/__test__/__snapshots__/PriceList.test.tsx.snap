// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshots 1`] = `ShallowWrapper {}`;
